src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/Pbkdf2Implementation.Browser.cs

```diff
@@ -0,0 +1,77 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using Internal.Cryptography;
+
+using SimpleDigest = Interop.BrowserCrypto.SimpleDigest;
+
+namespace System.Security.Cryptography
+{
+    internal static partial class Pbkdf2Implementation
+    {
+        public static void Fill(
+            ReadOnlySpan<byte> password,
+            ReadOnlySpan<byte> salt,
+            int iterations,
+            HashAlgorithmName hashAlgorithmName,
+            Span<byte> destination)
+        {
+            Debug.Assert(!destination.IsEmpty);
+            Debug.Assert(hashAlgorithmName.Name is not null);
+
+            if (Interop.BrowserCrypto.CanUseSubtleCrypto)
+            {
+                FillSubtleCrypto(password, salt, iterations, hashAlgorithmName, destination);
+            }
+            else
+            {
+                FillManaged(password, salt, iterations, hashAlgorithmName, destination);
+            }
+        }
+
+        private static unsafe void FillSubtleCrypto(
+            ReadOnlySpan<byte> password,
+            ReadOnlySpan<byte> salt,
+            int iterations,
+            HashAlgorithmName hashAlgorithmName,
+            Span<byte> destination)
+        {
+            (SimpleDigest hashName, _) = SHANativeHashProvider.HashAlgorithmToPal(hashAlgorithmName.Name!);
+
+            fixed (byte* pPassword = password)
+            fixed (byte* pSalt = salt)
+            fixed (byte* pDestination = destination)
+            {
+                int result = Interop.BrowserCrypto.DeriveBits(
+                    pPassword, password.Length,
+                    pSalt, salt.Length,
+                    iterations,
+                    hashName,
+                    pDestination, destination.Length);
+
+                if (result != 0)
+                    throw new CryptographicException(SR.Format(SR.Unknown_SubtleCrypto_Error, result));
+            }
+        }
+
+        private static void FillManaged(
+            ReadOnlySpan<byte> password,
+            ReadOnlySpan<byte> salt,
+            int iterations,
+            HashAlgorithmName hashAlgorithmName,
+            Span<byte> destination)
+        {
+            using (Rfc2898DeriveBytes deriveBytes = new Rfc2898DeriveBytes(
+                password.ToArray(),
+                salt.ToArray(),
```

Comment:
There's no way to stay in the span universe all the way down?  Are we accepting arrays from public entrypoints, passing them around as span, and then ToArray'ing here?

Reply:
Note that this is the way the existing "mobile" platforms are implementing this functionality:

https://github.com/dotnet/runtime/blob/6a02d5d3f35c84915bb71557569564ad07470238/src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/Pbkdf2Implementation.Managed.cs#L27-L36

which is used by Android and non-macOS AppleCrypto (iOS, tvOS, etc.).

> There's no way to stay in the span universe all the way down?

Not without some decent refactoring to `Rfc2898DeriveBytes` in order to make the managed algorithm completely "one shot", and then build the instance based `Rfc2898DeriveBytes` methods on top of it.

 > Are we accepting arrays from public entrypoints, passing them around as span, and then ToArray'ing here?

In one overload, yes, and in most overloads no. This method is called from these public end-points:

https://github.com/dotnet/runtime/blob/6a02d5d3f35c84915bb71557569564ad07470238/src/libraries/System.Security.Cryptography/ref/System.Security.Cryptography.cs#L1734-L1739

Only the top overload is taking 2 `byte[]`s, passing them through as spans, and then ToArray'ing here. The bottom overload takes a string, which turns into UTF8 bytes in either a rented array or stackalloc'd span, and then passed in here.

Reply:
> I have a branch with this. Happy to upstream it as a follow up to this.

I think that would be great!